### PR TITLE
Prevent duplicate transactions

### PR DESCRIPTION
**Pull request by StuAA78**
https://trello.com/c/FkpwxW8V/1866-prevent-duplicate-transactions-better-error-handling

WRLS have requested a change to the add transaction end point to help in situations where adding a transaction fails. When calling the endpoint, they will specify a `clientId`; if a transaction with this id doesn't already exist then a new transaction will be created, otherwise the transaction id will be returned. This means that a failed transaction can be retried without creating duplicate transactions.

This PR introduces a `find_transaction_by_client_id` service which can be given a client id and regime, and will return a matching transaction if one exists (or `null` if it doesn't). This is then used by the add transaction endpoint to check whether it's attempting to create a duplicate transaction and act accordingly.

### CODE REVIEW

File: test/services/find_transaction_by_client_id.test.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    schema = Schema[regime.slug]
+    region = 'A'
+    clientId = 'abc123'
+  })
+
+  beforeEach(async () => {
+    await cleanBillRuns()
+  })
+
+  it('returns the transaction id if a transaction with the client id exists', async () => {
```
Comment (Cruikshanks) [inline test/services/find_transaction_by_client_id.test.js:28]: Slight tweak to the wording to be more verbose about

- what gets returned from the service
- what we are actually matching against

```suggestion
  it('returns the matching transaction if one with the same regime and client id exists', async () => {
```

File: test/services/find_transaction_by_client_id.test.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    const billRun = await (schema.BillRun).find(regime.id, br.id)
+    const tId = await addBillRunTransaction(regime, billRun, { region: billRun.region, clientId })
+
+    const result = await FindTransactionByClientId.call(regime, clientId)
+
+    expect(result.id).to.equal(tId)
+    expect(result.clientId).to.equal(clientId)
+  })
+
+  it('returns null if no transaction with the client id exists', async () => {
```
Comment (Cruikshanks) [inline test/services/find_transaction_by_client_id.test.js:39]:
```suggestion
  it('returns null if no transaction with the same regime and client id exists', async () => {
```

**Review by Cruikshanks — CHANGES_REQUESTED** [review]
There's a typo that defo needs fixing.

I've suggested some updated titles for a couple of the tests but take them just as suggestions.

The key thing to look at is whether we can optimise the query to the database for a matching transaction and use its initial result rather than make a subsequent query. I'm interested to hear if I'm barking up the wrong tree!

File: app/services/find_transaction_by_client_id.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
+const { pool } = require('../lib/connectors/db')
+const Schema = require('../schema')
+
+async function call (regime, clientId) {
+  const stmt = 'select pre_sroc from transactions where client_id=$1 AND regime_id=$2::uuid'
```
[the remark this answers is not in the document]
Comment (StuAA78) [reply app/services/find_transaction_by_client_id.js]: Sounds good -- I've applied this change in the latest commit.

**Review by Cruikshanks — APPROVED** [review]

**Pull request merged**